### DISCUSSION TITLE
Google Coral Support

### DISCUSSION THREAD

**Discussion by tomtom215** [mmcc-xx/BirdCAGE#22] [Ideas]
Just an Idea:

I saw you mentioned you got at least a working version running on a Raspberry Pi 4, I think if you were able to offload any inferencing to a USB Coral TPU it would make it much lighter on CPU and RAM usage. This would also be a big differentiator over the BirdNetPi project IMO. It would also make sense if you were looking to marry this project with your other one which would require even more resources when running together.

https://coral.ai/docs/reference/py/

I would be happy to test when i'm available if you don't have one

**Comment by mmcc-xx** [comment]
I do happen to have one I'm using with Frigate. Do you know if It is possible for more than one container to use the TPU at the same time? I just assumed - without looking into it - that it would be bound to a specific container and that is it.

**Comment by mmcc-xx** [comment]
https://github.com/google-coral/edgetpu/issues/645

> Hi @aslanpour, TPU device is not designed to be (easily) shared between several processes or containers. As an option you can try to implement client/server architecture, i.e. server owns the TPU device and handles inference requests from multiple clients. Otherwise you need some mechanism to implement mutual exclusion between containers (e.g. you could try file system locks).

If this is correct, sounds like a hill for climbers.

**Reply by tomtom215** [reply]
> [google-coral/edgetpu#645](https://github.com/google-coral/edgetpu/issues/645)
> 
> > Hi @aslanpour, TPU device is not designed to be (easily) shared between several processes or containers. As an option you can try to implement client/server architecture, i.e. server owns the TPU device and handles inference requests from multiple clients. Otherwise you need some mechanism to implement mutual exclusion between containers (e.g. you could try file system locks).
> 
> If this is correct, sounds like a hill for climbers.

Oof, was unaware of this myself and I didn't think of this when I suggested it. I was considering a second Coral to leave one for Frigate and one for projects like this.

**Reply by mmcc-xx** [reply]
That might not be correct. I'm checking into it - just posted a question to the frigate subreddit. ChatGPT seems to think it is possible to just point multiple containers at the same TPU :)

**Reply by mmcc-xx** [reply]
According to someone who sounds like they know what they are talking about, only one process can interact with the TPU. Too bad.
